### PR TITLE
Change the logic of reinforcement of the guest hero's army during the castle defense

### LINKED ISSUE #6740: Combat screen, disposition of troops is not maintained when a hero is attacked in a castle and the garrisoned troops are transferred to the hero's army
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

In fheroes2, the disposition of troops is not maintained when a hero is attacked in a castle and the garrisoned troops are transferred to the hero's army.
All troops of the same type are combined into one troop.
This worked properly before.

In fheroes2:

Broken siege troop disposition 001

Broken siege troop disposition 002

### Save file

Here is the save file (just end the turn):
[Auto troop broken 01.zip](https://github.com/ihhub/fheroes2/files/10859844/Auto.troop.broken.01.zip)

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
close #6740

Hi @LeHerosInconnu could you please take a look at this implementation to find out whether it matches the expected logic?

### CODE REVIEW

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> Hi @LeHerosInconnu could you please take a look at this implementation to find out whether it matches the expected logic?

Yes, I'll look at it tomorrow.

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz,

> @oleg-derevenetz.
> 
> > Hi @LeHerosInconnu could you please take a look at this implementation to find out whether it matches the expected logic?
> 
> Yes, I'll look at it tomorrow.

I have tested all the arrangements presented as well as some others and it works as desired.
Good work. :)

**Comment by oleg-derevenetz** [comment]
OK, after some thought, I decided to do the following:

* For AI, defending hero will get the maximum possible army (i.e. slots in the hero army will be merged "horizontally" if necessary, and units from the garrison will attempt to join in descending order of their strength). In other words, the behavior will be the same as the "mainstream" behavior in other HoMM series games;
* For human player, no slots will be merged neither "horizontally" nor "vertically", units from garrison will attempt to join from left to right to exactly the same slots of the defending hero's army only if target slots are empty.

Hi @LeHerosInconnu @Branikolog @zenseii I invite you to look at the current implementation.

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz,

> OK, after some thought, I decided to do the following:
> 
> * For AI, defending hero will get the maximum possible army (i.e. slots in the hero army will be merged "horizontally" if necessary, and units from the garrison will attempt to join in descending order of their strength). In other words, the behavior will be the same as the "mainstream" behavior in other HoMM series games;
> * For human player, no slots will be merged neither "horizontally" nor "vertically", units from garrison will attempt to join from left to right to exactly the same slots of the defending hero's army only if target slots are empty.
> 
> Hi @LeHerosInconnu @Branikolog @zenseii I invite you to look at the current implementation.

I did some quick tests, everything seems to work as you described.
I will do a more thorough test in a full game later when I will have finished an already started game.

Have you made any changes to what is described below in this implementation?
"For the AI:
When an AI hero goes to spend the night in a castle, his army is reorganized with the garrisoned army to have the best possible army (also taking into account morale, etc.), this during the AI's turn.
When the AI hero leaves the castle the next turn, the armies are reorganized so that the original castle defense is not too reduced, or regains its original defense."

**Comment by oleg-derevenetz** [comment]
> Have you made any changes to what is described below in this implementation?

No, this pull request is only related to the pre-battle arrangement for the castle defense.

File: src/fheroes2/army/army.cpp [hunk 1]
```diff
@@ -1557,37 +1605,73 @@ void Army::resetInvalidMonsters() const
...
+    assert( commander != nullptr && commander->isHeroes() );
     // This method is designed to take reinforcements only from the garrison, because
     // it can leave the garrison empty
     assert( garrison.commander == nullptr || garrison.commander->isCaptain() );
 
-    // There are no troops in the garrison
-    if ( !garrison.isValid() ) {
-        return;
-    }
+    // If the guest hero's army is controlled by AI, then try to squeeze as much garrison troops as possible
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/army/army.cpp]: @zenseii thank you!

**Review by Branikolog — APPROVED** [review]
Everything works nice!
Thank you, @oleg-derevenetz , for this improvement! I believe, everyone will enjoy this new logic.

**Review by ihhub — APPROVED** [review]

**Pull request merged**